propagation/propagation.go

```diff
@@ -23,97 +45,159 @@ import (
 // required. If either is absent a nil trace header will be returned.
-func UnmarshalTraceContextV1(header string) (*Propagation, error) {
+func (HoneycombContext) extractV1(header string) (*trace.SpanContext, error) {
 	clauses := strings.Split(header, ",")
-	var prop = &Propagation{}
+	var sc = &trace.SpanContext{}
 	var tcB64 string
 	for _, clause := range clauses {
 		keyval := strings.SplitN(clause, "=", 2)
 		switch keyval[0] {
 		case "trace_id":
-			prop.TraceID = keyval[1]
+			sc.TraceID = keyval[1]
 		case "parent_id":
-			prop.ParentID = keyval[1]
+			sc.ParentID = keyval[1]
 		case "dataset":
-			prop.Dataset, _ = url.QueryUnescape(keyval[1])
+			sc.Dataset, _ = url.QueryUnescape(keyval[1])
 		case "context":
 			tcB64 = keyval[1]
 		}
 	}
-	if prop.TraceID == "" && prop.ParentID != "" {
-		return nil, &PropagationError{"parent_id without trace_id", nil}
+	if sc.TraceID == "" && sc.ParentID != "" {
+		return nil, &propagationError{"parent_id without trace_id", nil}
 	}
 	if tcB64 != "" {
 		data, err := base64.StdEncoding.DecodeString(tcB64)
 		if err != nil {
-			return nil, &PropagationError{"unable to decode base64 trace context", err}
+			return nil, &propagationError{"unable to decode base64 trace context", err}
 		}
-		prop.TraceContext = make(map[string]interface{})
-		err = json.Unmarshal(data, &prop.TraceContext)
+		sc.TraceContext = make(map[string]interface{})
+		err = json.Unmarshal(data, &sc.TraceContext)
 		if err != nil {
-			return nil, &PropagationError{"unable to unmarshal trace context", err}
+			return nil, &propagationError{"unable to unmarshal trace context", err}
 		}
 	}
-	return prop, nil
+	return sc, nil
+
+}
+
+// Inject assembles the trace context header and sets it in the supplier.
+func (HoneycombContext) Inject(ctx context.Context, supplier otelprop.HTTPSupplier) {
+	sc := trace.GetRemoteSpanContextFromContext(ctx)
+	if sc == nil {
+		return
+	}
+	h := sc.SerializeHeader()
+	supplier.Set(honeycombTracePropagationHTTPHeader, h)
+}
+
+// GetAllKeys returns the name of the header used for trace context propagation.
+func (HoneycombContext) GetAllKeys() []string {
+	return []string{honeycombTracePropagationHTTPHeader}
+}
+
+// AmazonContext understands how to parse and generate Amazon trace propagation headers
+type AmazonContext struct{}
+
+// AmazonHTTPPropagator returns an HTTPPropagator that understands Amazon trace context
+// headers.
+func AmazonHTTPPropagator() AmazonContext {
+	return AmazonContext{}
+}
+
+// Extract takes the trace header and creates a SpanContext which is then stored in the
+// provided context.
+func (AmazonContext) Extract(ctx context.Context, supplier otelprop.HTTPSupplier) context.Context {
+	header := supplier.Get(amazonTracePropagationHTTPHeader)
+	segments := strings.Split(header, ";")
+
+	sc := &trace.SpanContext{}
+	for _, segment := range segments {
+		keyval := strings.SplitN(segment, "=", 2)
+		switch strings.ToLower(keyval[0]) {
+		case "self":
```

Comment:
Doesn't the Amazon header use `Root` and `Parent`, as in the `Extract` code below? Should these be defined constants?

Reply:
Amazon Load Balancers use `Root` and `Self`:

https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-request-tracing.html

I should include support for instrumented applications though, and parse `Parent`. It also occurs to me that additional fields are allowed, so we can put those in the `TraceContext` field.